File: backend/src/main/java/com/example/knowledgehub/backend/category/CategoryDtos.java
package com.example.knowledgehub.backend.category;

import jakarta.validation.constraints.NotBlank;

public class CategoryDtos {

    public record CategoryResponse(
            Long id,
            String code,
            String name,
            String description,
            String teacherName,
            String teacherEmail,
            String teacherPhone,
            String teacherLinkedInUrl
    ) {
    }

    public record CategoryCreateRequest(
            @NotBlank String code,
            @NotBlank String name,
            String description,
            String teacherName,
            String teacherEmail,
            String teacherPhone,
            String teacherLinkedInUrl
    ) {
    }

    public record CategoryUpdateRequest(
            String code,
            String name,
            String description,
            String teacherName,
            String teacherEmail,
            String teacherPhone,
            String teacherLinkedInUrl
    ) {
    }

    public static CategoryResponse toResponse(Category entity) {
        if (entity == null) return null;
        return new CategoryResponse(
                entity.getId(),
                entity.getCode(),
                entity.getName(),
                entity.getDescription(),
                entity.getTeacherName(),
                entity.getTeacherEmail(),
                entity.getTeacherPhone(),
                entity.getTeacherLinkedInUrl()
        );
    }

    public static Category fromCreateRequest(CategoryCreateRequest request) {
        if (request == null) return null;
        return new Category(
                request.code(),
                request.name(),
                request.description(),
                request.teacherName(),
                request.teacherEmail(),
                request.teacherPhone(),
                request.teacherLinkedInUrl()
        );
    }

    public static void applyUpdate(Category entity, CategoryUpdateRequest update) {
        if (entity == null || update == null) return;
        if (update.code() != null) entity.setCode(update.code());
        if (update.name() != null) entity.setName(update.name());
        if (update.description() != null) entity.setDescription(update.description());
        if (update.teacherName() != null) entity.setTeacherName(update.teacherName());
        if (update.teacherEmail() != null) entity.setTeacherEmail(update.teacherEmail());
        if (update.teacherPhone() != null) entity.setTeacherPhone(update.teacherPhone());
        if (update.teacherLinkedInUrl() != null) entity.setTeacherLinkedInUrl(update.teacherLinkedInUrl());
    }
}

File: backend/src/main/java/com/example/knowledgehub/backend/category/CategoryDtos.kt
package com.example.knowledgehub.backend.category

import jakarta.validation.constraints.NotBlank

data class CategoryResponse(
    val id: Long?,
    val code: String,
    val name: String,
    val description: String?,
    val teacherName: String?,
    val teacherEmail: String?,
    val teacherPhone: String?,
    val teacherLinkedInUrl: String?,
)

data class CategoryCreateRequest(
    @field:NotBlank
    val code: String,

    @field:NotBlank
    val name: String,
    val description: String? = null,
    val teacherName: String? = null,
    val teacherEmail: String? = null,
    val teacherPhone: String? = null,
    val teacherLinkedInUrl: String? = null,
)

data class CategoryUpdateRequest(
    val code: String? = null,
    val name: String? = null,
    val description: String? = null,
    val teacherName: String? = null,
    val teacherEmail: String? = null,
    val teacherPhone: String? = null,
    val teacherLinkedInUrl: String? = null,
)

fun Category.toResponse() = CategoryResponse(
    id = id,
    code = code,
    name = name,
    description = description,
    teacherName = teacherName,
    teacherEmail = teacherEmail,
    teacherPhone = teacherPhone,
    teacherLinkedInUrl = teacherLinkedInUrl,
)

fun CategoryCreateRequest.toEntity() = Category(
    code = code,
    name = name,
    description = description,
    teacherName = teacherName,
    teacherEmail = teacherEmail,
    teacherPhone = teacherPhone,
    teacherLinkedInUrl = teacherLinkedInUrl,
)

fun Category.applyUpdate(update: CategoryUpdateRequest) {
    update.code?.let { this.code = it }
    update.name?.let { this.name = it }
    if (update.description != null) this.description = update.description
    if (update.teacherName != null) this.teacherName = update.teacherName
    if (update.teacherEmail != null) this.teacherEmail = update.teacherEmail
    if (update.teacherPhone != null) this.teacherPhone = update.teacherPhone
    if (update.teacherLinkedInUrl != null) this.teacherLinkedInUrl = update.teacherLinkedInUrl
}
